File: rotateimage.js
//Given a 2-D matrix, rotate it 90 degrees clockwise in-place.

let matrix = 
[[1,2,3],
 [4,5,6],
 [7,8,9]]


//O(n^2) solution that transposes the matrix first then reverses the rows.

//Transpose the matrix; break if you pass by [1,1], [2,2], etc.
for (let i = 0; i < matrix.length; i++) {
    for (let j = 0; j < matrix.length; j++) {
        if (i == j) {
            break
        }
        let temp = matrix[i][j]
        matrix[i][j] = matrix[j][i]
        matrix[j][i] = temp
    }
}

//Reverse each row
for (let i = 0; i < matrix.length; i++) {
    matrix[i] = matrix[i].reverse()
}

File: rotateimage.java
//Java Solution

class Solution {
    public void rotate(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                if (i == j) {
                    break;
                }
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }
        
        for (int[] row : matrix) {
            row = reverse(row);
        }
    }
    
    public int[] reverse(int[] row) {
        int left = 0;
        int right = row.length - 1;
        
        while (left < right) {
            int temp = row[left];
            row[left++] = row[right];
            row[right--] = temp;
        }
        
        return row;
    }
}